fix: scan entries when the root is in node_modules

Jinjiang opened the pull request:
<!-- Thank you for contributing! -->

### Description

Fixed optimizer entries in the case the root is in node_modules

### Additional context

https://github.com/vitejs/vite/discussions/15613#discussioncomment-8277522

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md), especially the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Update the corresponding documentation if needed.
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

Jinjiang committed:
fix: scan entries when the root is in node_modules

patak-dev commented on packages/vite/src/node/optimizer/scan.ts:
```diff
@@ -242,7 +242,7 @@ function globEntries(pattern: string | string[], config: ResolvedConfig) {
     cwd: config.root,
     ignore: [
-      '**/node_modules/**',
+      glob.convertPathToPattern(config.root) + '/**/node_modules/**',
```
This shouldn't be needed. Could you explain why this is solving the issue? We are doing the glob with `cwd` at `config.root`, so if passing absolute paths to ignore is supported, it should give the same result as the one we already have here. Or there is a bug in `glob` or maybe passing this new pattern has the effect of invalidating the ignore for all users, so your app starts to work.

Jinjiang replied:
Sure. Let me explain:

Let's say the full absolute path of `index.html` is: `/<my-workspace>/node_modules/@bitdev/react.examples.button/index.html`

Then the original `**/node_modules/**` will match this path and make it ignored by `fast-glob`.

However, `/<my-workspace>/node_modules/@bitdev/react.examples.button/**/node_modules/**` won't match it and it works.

Below is the log (before and after this PR):

Below:

```
input: {
  pattern: '**/*.html',
  cwd: '/Volumes/jinjiang-portable-disk/Developer/personal/reproductions/node_modules/.pnpm/@bitdev+react.examples.button@0.0.23_@testing-library+react@12.1.5_@types+react@18.2.48_react_5npp46vlhu2o2i5ubj23gk35sy/node_modules/@bitdev/react.examples.button',
  ignore: [
    '**/node_modules/**',
    '**/dist/**',
    '**/__tests__/**',
    '**/coverage/**'
  ],
  absolute: true,
  suppressErrors: true
}
result: { entries: [] }
```


After:

```
input: {
  pattern: '**/*.html',
  cwd: '/Volumes/jinjiang-portable-disk/Developer/personal/reproductions/node_modules/.pnpm/@bitdev+react.examples.button@0.0.23_@testing-library+react@12.1.5_@types+react@18.2.48_react_5npp46vlhu2o2i5ubj23gk35sy/node_modules/@bitdev/react.examples.button',
  ignore: [
    '/Volumes/jinjiang-portable-disk/Developer/personal/reproductions/node_modules/.pnpm/@bitdev+react.examples.button@0.0.23_@testing-library+react@12.1.5_@types+react@18.2.48_react_5npp46vlhu2o2i5ubj23gk35sy/node_modules/@bitdev/react.examples.button/**/node_modules/**',
    '**/dist/**',
    '**/__tests__/**',
    '**/coverage/**'
  ],
  absolute: true,
  suppressErrors: true
}
result: {
  entries: [
    '/Volumes/jinjiang-portable-disk/Developer/personal/reproductions/node_modules/.pnpm/@bitdev+react.examples.button@0.0.23_@testing-library+react@12.1.5_@types+react@18.2.48_react_5npp46vlhu2o2i5ubj23gk35sy/node_modules/@bitdev/react.examples.button/index.html'
  ]
}
```

patak-dev replied:
Ok, I see 👍🏼

patak-dev commented:
I think we could merge this PR. We should do the same for `**/${config.build.outDir}/**`, in that case, so we also avoid potential problems if `outDir` is part of the parent too. And the same for `**/__tests__/**` and `**/__coverage__/**` just to be consistent.

I'd like to see what other maintainers think about this too. Because even if your reproduction is working after this change, it may later fail unexpectedly because as I [commented here](https://github.com/vitejs/vite/discussions/15613#discussioncomment-8145383), our current `isInNodeModules` util does a simple `.includes('node_modules')`. So it may not make sense to merge this PR until we have proper tests for an app inside `node_modules` and this util has been reworked.

There is also a higher level discussion about the extra complexity we are introducing to support apps inside `node_modules`. I understand that configuring this in Bit isn't possible though so if the code changes aren't hard to maintain I'm leaning towards supporting this use case in core. I think you'll have a lot of issues in the ecosystem though, because there are a lot of plugins out there that have the same kind of checks to decide if something is a dependency file or not. It could be an uphill battle. So I would still recommend to check if there is any chance to configure Bit in a way were the Apps parent folder isn't `node_modules`.

Jinjiang commented:
@patak-dev thanks for the information. Ya, I agree in a larger picture there might be somewhere else I'm not familiar with having the similar potential ones.

On Bit's side, in short term, I'm afraid it's hard to change since the concept in Bit is that everything runs as a package, so inevitably the runtime happens in a node_modules directory. For other libs/plugins in the ecosystem, we also did some related work. By far, this issue is the last one. 😅 

In the long-term, I'd like to bring this topic to the team. /cc @GiladShoham

Thanks.

Jinjiang commented:
@patak-dev I've updated the code to cover other ignored patterns in `globEntries()`.

However, I met a CI error on Windows said `ResourceUnavailable: D:\a\_temp\9a25bc22-181a-46f2-b736-ec7ee0a6aa41.ps1:3`. I don't quite understand this error and it seems not related to what I've changed. Do you have any ideas on this? Neither not sure a re-run would help for which I don't have that permission.

🙏

patak-dev commented:
Don't worry about Windows, it seems it is failing across the board. I think we could merge this one, but explicitly stating that what you are doing isn't supported. Let's wait for @bluwy that said wanted to take a look

bluwy reviewed (approved):
Looks fine to me 👍 I thought this was touching the watcher ignores, which I'll have more reservations of if so, but since this is only for the entries, it looks like a good change for now.

Jinjiang commented:
Thanks so much both.

So I guess we can merge this without the Windows CI pass? 🤔 
Let me know anything I can help further.

patak-dev commented:
There is an issue with playwright across the board, I'll merge this one without that test. Thanks for the patience @Jinjiang!
